File: backend/crud/cached.py
"""
Cached CRUD operations for performance optimization.

This module provides cached versions of frequently called CRUD operations
to reduce database load, especially for polling endpoints.

Caching strategy:
- Agents: 5 minute TTL (rarely change)
- Rooms: 30 second TTL (max_interactions, is_paused may change)
- Room agents: 1 minute TTL (agents added/removed occasionally)
- Messages: 5 second TTL (high write frequency, need fresh data)

IMPORTANT: SQLAlchemy ORM objects are detached from their session before caching
using make_transient() to prevent DetachedInstanceError when accessed later.
"""

import logging
from typing import List, Optional

from infrastructure.cache import agent_object_key, get_cache, room_agents_key, room_messages_key, room_object_key
from infrastructure.database import models
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import make_transient

import crud

logger = logging.getLogger("CachedCRUD")


async def get_agent_cached(db: AsyncSession, agent_id: int) -> Optional[models.Agent]:
    """
    Get agent with caching (TTL: 5 minutes).

    The agent object is detached from the session using make_transient()
    to prevent DetachedInstanceError when accessed from cache later.

    Args:
        db: Database session
        agent_id: Agent ID

    Returns:
        Agent object or None
    """
    cache = get_cache()
    key = agent_object_key(agent_id)

    async def fetch_and_detach():
        agent = await crud.get_agent(db, agent_id)
        if agent:
            make_transient(agent)
        return agent

    return await cache.get_or_set_async(
        key=key,
        factory=fetch_and_detach,
        ttl_seconds=300,  # 5 minutes
    )


async def get_room_cached(db: AsyncSession, room_id: int) -> Optional[models.Room]:
    """
    Get room with caching (TTL: 30 seconds).

    The room object is detached from the session using make_transient()
    to prevent DetachedInstanceError when accessed from cache later.

    Args:
        db: Database session
        room_id: Room ID

    Returns:
        Room object or None
    """
    cache = get_cache()
    key = room_object_key(room_id)

    async def fetch_and_detach():
        room = await crud.get_room(db, room_id)
        if room:
            # Detach from session to prevent DetachedInstanceError when cached
            make_transient(room)
            # Also detach related agents to prevent lazy-load issues
            for agent in room.agents:
                make_transient(agent)
            # Detach world if loaded (for TRPG rooms)
            if room.world:
                make_transient(room.world)
        return room

    return await cache.get_or_set_async(
        key=key,
        factory=fetch_and_detach,
        ttl_seconds=30,  # 30 seconds
    )


async def get_agents_cached(db: AsyncSession, room_id: int) -> List[models.Agent]:
    """
    Get agents in a room with caching (TTL: 1 minute).

    Agent objects are detached from the session using make_transient()
    to prevent DetachedInstanceError when accessed from cache later.

    Args:
        db: Database session
        room_id: Room ID

    Returns:
        List of agents in the room
    """
    cache = get_cache()
    key = room_agents_key(room_id)

    async def fetch_and_detach():
        agents = await crud.get_agents(db, room_id)
        for agent in agents:
            make_transient(agent)
        return agents

    return await cache.get_or_set_async(
        key=key,
        factory=fetch_and_detach,
        ttl_seconds=60,  # 1 minute
    )


async def get_messages_cached(db: AsyncSession, room_id: int) -> List[models.Message]:
    """
    Get messages in a room with caching (TTL: 5 seconds).

    Note: Short TTL because messages are written frequently.
    For polling, use get_messages_since which doesn't need caching.

    Message objects are detached from the session using make_transient()
    to prevent DetachedInstanceError when accessed from cache later.

    Args:
        db: Database session
        room_id: Room ID

    Returns:
        List of all messages in the room
    """
    cache = get_cache()
    key = room_messages_key(room_id)

    async def fetch_and_detach():
        messages = await crud.get_messages(db, room_id)
        for msg in messages:
            make_transient(msg)
            if msg.agent:
                make_transient(msg.agent)
        return messages

    return await cache.get_or_set_async(
        key=key,
        factory=fetch_and_detach,
        ttl_seconds=5,  # 5 seconds
    )


async def get_recent_messages_cached(db: AsyncSession, room_id: int, limit: int = 200) -> List[models.Message]:
    """
    Get recent messages with caching to avoid repeated DB hits.

    Message objects are detached from the session using make_transient()
    to prevent DetachedInstanceError when accessed from cache later.
    """
    cache = get_cache()
    key = f"{room_messages_key(room_id)}:recent:{limit}"

    async def fetch_and_detach():
        messages = await crud.get_recent_messages(db, room_id, limit)
        for msg in messages:
            make_transient(msg)
            if msg.agent:
                make_transient(msg.agent)
        return messages

    return await cache.get_or_set_async(
        key=key,
        factory=fetch_and_detach,
        ttl_seconds=5,
    )


async def get_messages_since_cached(
    db: AsyncSession,
    room_id: int,
    since_id: int | None = None,
    limit: int = 100,
) -> List[models.Message]:
    """
    Get new messages since a specific ID using a cached recent slice.

    This avoids a fresh database query on every poll while keeping the
    result bounded.
    """
    # Use a buffer to reduce cache misses when since_id increments rapidly
    buffered_limit = max(limit * 2, 50)
    recent_messages = await get_recent_messages_cached(db, room_id, buffered_limit)

    if since_id is None:
        return recent_messages[:limit]

    filtered = [m for m in recent_messages if m.id > since_id]
    return filtered[:limit]


async def get_messages_after_agent_response_cached(
    db: AsyncSession,
    room_id: int,
    agent_id: int,
    limit: int = 200,
) -> List[models.Message]:
    """
    Get messages after an agent's last response, cached briefly.

    Message objects are detached from the session using make_transient()
    to prevent DetachedInstanceError when accessed from cache later.
    """
    cache = get_cache()
    key = f"{room_messages_key(room_id)}:after:{agent_id}:{limit}"

    async def fetch_and_detach():
        messages = await crud.get_messages_after_agent_response(db, room_id, agent_id, limit)
        for msg in messages:
            make_transient(msg)
            if msg.agent:
                make_transient(msg.agent)
        return messages

    return await cache.get_or_set_async(
        key=key,
        factory=fetch_and_detach,
        ttl_seconds=5,
    )


# Cache invalidation helpers


def invalidate_room_cache(room_id: int):
    """
    Invalidate all cache entries related to a room.

    Args:
        room_id: Room ID
    """
    cache = get_cache()
    cache.invalidate(room_object_key(room_id))
    cache.invalidate(room_agents_key(room_id))
    # Invalidate all message-related entries (full list + recent slices)
    cache.invalidate_pattern(room_messages_key(room_id))
    logger.debug(f"Invalidated cache for room {room_id}")


def invalidate_agent_cache(agent_id: int):
    """
    Invalidate all cache entries related to an agent.

    Args:
        agent_id: Agent ID
    """
    from infrastructure.cache import agent_config_key

    cache = get_cache()
    cache.invalidate(agent_object_key(agent_id))
    cache.invalidate(agent_config_key(agent_id))
    logger.debug(f"Invalidated cache for agent {agent_id}")


def invalidate_messages_cache(room_id: int):
    """
    Invalidate message cache for a room.

    Args:
        room_id: Room ID
    """
    cache = get_cache()
    cache.invalidate_pattern(room_messages_key(room_id))
    logger.debug(f"Invalidated message cache for room {room_id}")

File: backend/src/crud/cached.ts
/**
 * Cached reads for the CRUD the polling endpoints hammer. Sync `getOrSet`, not
 * `getOrSetAsync`: `bun:sqlite` runs a statement to completion before any other
 * JS, so there is no await window between miss and store. **Invalidation lives
 * at the writes** — the sweeps in `cache-invalidation.ts` plus inline sweeps in `rooms.ts`,
 * `messages.ts` and `agents.ts`; even the deletes sweep, since SQLite reuses
 * rowids. Only `deleteRoomMessages` skips it.
 */

import { getAgent } from './agents'
import type { Db } from '../db'
import type { Agent } from '../db/schema'
import {
  agentObjectKey,
  getCache,
  roomAgentsKey,
  roomMessagesKey,
  roomObjectKey,
} from '../infrastructure/cache'
import {
  getMessages,
  getMessagesAfterAgentResponse,
  getRecentMessages,
  type MessageWithAgent,
} from './messages'
import { getAgentsInRoom, getRoom, type RoomWithRelations } from './rooms'

// How stale a reader can stand to be: agent rows change only on a config edit,
// room flags gate the scheduler, a poller showing a turn late is user-visible.
const AGENT_TTL_SECONDS = 300
const ROOM_TTL_SECONDS = 30
const ROOM_AGENTS_TTL_SECONDS = 60
const MESSAGES_TTL_SECONDS = 5

/**
 * The long TTL is safe only because `invalidateAgentCache` (in `cache-invalidation.ts`) runs on every
 * `updateAgent`; without it a config hot-reload would take 5 minutes to land.
 */
export function getAgentCached(db: Db, agentId: number): Agent | null {
  return getCache().getOrSet<Agent | null>(
    agentObjectKey(agentId),
    () => getAgent(db, agentId),
    AGENT_TTL_SECONDS,
  )
}

export function getRoomCached(db: Db, roomId: number): RoomWithRelations | null {
  return getCache().getOrSet<RoomWithRelations | null>(
    roomObjectKey(roomId),
    () => getRoom(db, roomId),
    ROOM_TTL_SECONDS,
  )
}

export function getAgentsCached(db: Db, roomId: number): Agent[] {
  return getCache().getOrSet<Agent[]>(
    roomAgentsKey(roomId),
    () => getAgentsInRoom(db, roomId),
    ROOM_AGENTS_TTL_SECONDS,
  )
}

/**
 * The key is `roomMessagesKey(roomId)` itself — both an entry and the prefix
 * every other message entry extends, so one sweep clears the family. Unbounded.
 */
export function getMessagesCached(db: Db, roomId: number): MessageWithAgent[] {
  return getCache().getOrSet<MessageWithAgent[]>(
    roomMessagesKey(roomId),
    () => getMessages(db, roomId),
    MESSAGES_TTL_SECONDS,
  )
}

export function getRecentMessagesCached(
  db: Db,
  roomId: number,
  limit = 200,
): MessageWithAgent[] {
  const key = `${roomMessagesKey(roomId)}:recent:${limit}`
  return getCache().getOrSet<MessageWithAgent[]>(
    key,
    () => getRecentMessages(db, roomId, limit),
    MESSAGES_TTL_SECONDS,
  )
}

/**
 * **Different semantics from the uncached `getMessagesSince`,** which returns
 * the *oldest* rows above `sinceId` so a client catches up page by page; this
 * filters the *newest* `bufferedLimit` rows, so a client further behind never
 * sees the rows in between. The buffer is why caching works — `sinceId` advances
 * every poll, so keying on it would miss.
 */
export function getMessagesSinceCached(
  db: Db,
  roomId: number,
  sinceId: number | null = null,
  limit = 100,
): MessageWithAgent[] {
  const bufferedLimit = Math.max(limit * 2, 50)
  const recent = getRecentMessagesCached(db, roomId, bufferedLimit)

  if (sinceId === null) return recent.slice(0, limit)

  return recent.filter((m) => m.id > sinceId).slice(0, limit)
}

/**
 * The key derives from `roomMessagesKey`, which is load-bearing: sharing that
 * prefix lets one sweep drop every message-derived entry for a room.
 */
export function getMessagesAfterAgentResponseCached(
  db: Db,
  roomId: number,
  agentId: number,
  limit = 200,
): MessageWithAgent[] {
  const key = `${roomMessagesKey(roomId)}:after:${agentId}:${limit}`
  return getCache().getOrSet<MessageWithAgent[]>(
    key,
    () => getMessagesAfterAgentResponse(db, roomId, agentId, limit),
    MESSAGES_TTL_SECONDS,
  )
}
